core/src/main/java/hudson/cli/listeners/CliListener.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2025, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package hudson.cli.listeners;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.ExtensionPoint;
+import hudson.cli.CLICommand;
+import jenkins.util.Listeners;
+
+/**
+ * Allows implementations to listen to {@link CLICommand#run()} execution events.
+ *
+ * @since TODO
+ */
+public interface CliListener extends ExtensionPoint {
+
+    /**
+     * Invoked before command execution.
+     *
+     * @param context Information about the command being executed
+     * */
+    default void onExecution(@NonNull CliContext context) {}
+
+    /**
+     * Invoked after successful execution.
+     *
+     * @param context Information about the command being executed
+     * @param exitCode `run` returned exit code.
+     * */
+    default void onCompleted(@NonNull CliContext context, int exitCode) {}
+
+    /**
+     * Invoked on execution failure.
+     *
+     * @param context Information about the command being executed
+     * @param exitCode `run` returned exit code.
```

Comment:
It seems unnecessary to pass this parameter to the listener. If anyone really cared it could be inferred by the same logic as in `CLICommand` but generally a listener is going to want to report any error at the Java level, not by using a Unix-like exit code value.

Reply:
ATM exit code is used to respect the current log levels. Notably unhanded exceptions and BadCredentials.
We can indeed check the same kind of Exceptions as in CLICommand#handleException, but  exit codes handled by jenkins are properly documented on CLICommand#main, perhaps more maintainable ?

Reply:
exit code removed in the signature with the exception